internal/coreinternal/golden/sort_metrics.go

```diff
@@ -85,3 +87,75 @@ func sortAttributeMap(mp pcommon.Map) {
 	}
 	tempMap.CopyTo(mp)
 }
+
+// sortMetricDataPointSlices sorts the datapoint slice of a pmetric.Metrics according to the alphanumeric ordering of map key
+func sortMetricDataPointSlices(ms pmetric.Metrics) {
+	for i := 0; i < ms.ResourceMetrics().Len(); i++ {
+		for j := 0; j < ms.ResourceMetrics().At(i).ScopeMetrics().Len(); j++ {
+			for k := 0; k < ms.ResourceMetrics().At(i).ScopeMetrics().At(j).Metrics().Len(); k++ {
+				m := ms.ResourceMetrics().At(i).ScopeMetrics().At(j).Metrics().At(k)
+				switch m.Type() {
+				case pmetric.MetricTypeGauge:
+					sortNumberDataPointSlice(m.Gauge().DataPoints())
+				case pmetric.MetricTypeSum:
+					sortNumberDataPointSlice(m.Sum().DataPoints())
+				case pmetric.MetricTypeHistogram:
+					sortHistogramDataPointSlice(m.Histogram().DataPoints())
+				case pmetric.MetricTypeExponentialHistogram:
+					sortExponentialHistogramDataPointSlice(m.ExponentialHistogram().DataPoints())
+				case pmetric.MetricTypeSummary:
+					sortSummaryDataPointSlice(m.Summary().DataPoints())
+				}
+			}
+		}
+	}
+}
+
+func sortNumberDataPointSlice(ndps pmetric.NumberDataPointSlice) {
+	ndps.Sort(func(a, b pmetric.NumberDataPoint) bool {
+		return compareMaps(a.Attributes(), b.Attributes()) < 0
+	})
+}
+
+func sortSummaryDataPointSlice(sdps pmetric.SummaryDataPointSlice) {
+	sdps.Sort(func(a, b pmetric.SummaryDataPoint) bool {
+		return compareMaps(a.Attributes(), b.Attributes()) < 0
+	})
+}
+
+func sortHistogramDataPointSlice(hdps pmetric.HistogramDataPointSlice) {
+	hdps.Sort(func(a, b pmetric.HistogramDataPoint) bool {
+		return compareMaps(a.Attributes(), b.Attributes()) < 0
+	})
+}
+
+func sortExponentialHistogramDataPointSlice(ehdps pmetric.ExponentialHistogramDataPointSlice) {
+	ehdps.Sort(func(a, b pmetric.ExponentialHistogramDataPoint) bool {
+		return compareMaps(a.Attributes(), b.Attributes()) < 0
+	})
+}
+
+// assumes the parameters have been pre-sorted
+func compareMaps(a, b pcommon.Map) int {
+	if a.Len() != b.Len() {
+		return a.Len() - b.Len()
+	}
+
+	var aKeys, bKeys []string
+	a.Range(func(k string, _ pcommon.Value) bool {
+		aKeys = append(aKeys, k)
+		return true
+	})
+	b.Range(func(k string, _ pcommon.Value) bool {
+		bKeys = append(bKeys, k)
+		return true
+	})
+
+	for i := 0; i < len(aKeys); i++ {
+		if aKeys[i] != bKeys[i] {
+			return strings.Compare(aKeys[i], bKeys[i])
```

Comment:
Don't we need to sort `aKeys` and `bKeys` before comparing elements?

Reply:
This takes advantage of the fact that the attribute keys should have been sorted previously by the work done in the previous MR where we sort the attributes. The sort datapoint slice is called after that main sort where all the attributes are sorted so it should ensure these are always presorted.